consensus/src/liveness/proposer_election.rs

```diff
@@ -18,12 +20,45 @@ pub trait ProposerElection {
     fn get_valid_proposer(&self, round: Round) -> Author;
 }
 
-// next continuously mutates a state and returns a u64-index
-pub(crate) fn next(state: &mut Vec<u8>) -> u64 {
-    // state = SHA-3-256(state)
-    *state = aptos_crypto::HashValue::sha3_256_of(state).to_vec();
+// next consumes seed and returns random deterministic u64 value in [0, max) range
+fn next_in_range(state: Vec<u8>, max: u64) -> u64 {
+    // hash = SHA-3-256(state)
+    let hash = aptos_crypto::HashValue::sha3_256_of(&state).to_vec();
     let mut temp = [0u8; 8];
-    copy_slice_to_vec(&state[..8], &mut temp).expect("next failed");
-    // return state[0..8]
-    u64::from_le_bytes(temp)
+    copy_slice_to_vec(&hash[..8], &mut temp).expect("next failed");
+    // return hash[0..8]
+    u64::from_le_bytes(temp) % max
+}
+
+// chose index randomly, with given weight distribution
+pub(crate) fn choose_index(mut weights: Vec<u64>, state: Vec<u8>) -> usize {
+    let mut total_weight = 0;
+    for w in &mut weights {
```

Comment:
Its a bit confusing to reuse the weight vector and transform that to cumulative weight. Let's may be create a new vector instead of mutating in place ?

Reply:
since we own the vector, confusion should only be scoped to the function itself. adding the comment to explain it (instead of unnecessary allocation